### PR TITLE
readd toolbar

### PR DESCRIPTION
**Pull request by DerMolly**
### Component/Part
editor

### Description
This PR readds the toolbar for the editor.
### Steps

- [x] added implementation
- [x] added / updated tests
- [x] added / updated documentation
- [x] extended changelog

### CODE REVIEW

**Comment by DerMolly** [comment]
currently the toolbar does not wrap if the editor-window get's resized too small, that should be fixed before merging this

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
+import CodeMirror from 'codemirror'
+
+export const updateSelection = (content: string, startPosition: CodeMirror.Position, endPosition: CodeMirror.Position, onContentChange: (content: string) => void, selection: string): void => {
+  const lines = content.split('\n')
+  const newLines = selection.split('\n')
+  const extraLines = newLines.length - 1
```
Comment (mrdrogdrog) [inline src/components/editor/editor-window/tool-bar/utils.ts:6]: This hole file is very complicated and most of the variable names are confusing. 

Why `lines`? Why does `newLines` contain the selection? Why does `extraLines` contain a number and not line?

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
+import CodeMirror from 'codemirror'
+
```
Comment (mrdrogdrog) [inline src/components/editor/editor-window/tool-bar/utils.ts:2]: Did you know there is an array function `splice` that can delete items from an array AND insert items into the array at specific indexes? This and the usage of map and filter could make the content a lot easier to understand.

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    } else {
+      for (let i = lines.length - 1; i > startPosition.line; i--) {
+        lines[i + extraLines] = lines[i]
+      }
+      const lastPart = lines[startPosition.line].slice(endPosition.ch)
+      lines[startPosition.line] = lines[startPosition.line].slice(0, startPosition.ch) + newLines[0]
+      let j = 1
+      for (let i = 0; i < extraLines; i++) {
+        lines[i + startPosition.line + 1] = newLines[j]
+        j++
```
Comment (mrdrogdrog) [inline src/components/editor/editor-window/tool-bar/utils.ts:20]: why j ? j is just i + 1

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  }
+  if (startPosition.line !== endPosition.line && startPosition.ch !== endPosition.ch) {
+    for (let i = lines.length - 1; i > endPosition.line; i--) {
+      lines[i + extraLines] = lines[i]
+    }
+    let j = 0
+    for (let i = 0; i <= extraLines; i++) {
+      lines[i + startPosition.line] = newLines[j]
+      j++
+    }
```
Comment (mrdrogdrog) [inline src/components/editor/editor-window/tool-bar/utils.ts:33]: why j ? j = i

File: src/components/editor/editor-window/tool-bar/utils.ts (hunk 2 above)
Comment (mrdrogdrog) [reply src/components/editor/editor-window/tool-bar/utils.ts:2]: I stoped looking through this file after the `j = i` comment. Please clean this up first. :/

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
+import CodeMirror from 'codemirror'
+
+export const updateSelection = (content: string, startPosition: CodeMirror.Position, endPosition: CodeMirror.Position, onContentChange: (content: string) => void, selection: string): void => {
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts:3]: yeah, that's a good suggestion

File: src/components/editor/editor-window/tool-bar/utils.ts (hunk 1 above)
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts:6]: lines: the lines in the content
newLines: the lines in the selection (the new lines to add)
extraLines: this maybe should be numberOfExtraLines

File: src/components/editor/editor-window/tool-bar/utils.ts (hunk 2 above)
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts:2]: I just have a look at the splice functions and as far as the documentation tells me this helps with nothing at all

File: src/components/editor/editor-window/tool-bar/utils.ts (hunk 3 above)
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts:20]: it was needed once and then got optimized, but not deleted

File: src/components/editor/editor-window/tool-bar/utils.ts (hunk 4 above)
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts:33]: it was needed once and then got optimized, but not deleted

File: src/components/editor/editor-window/tool-bar/tool-bar.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  const makeSelectionItalic = () => addMarkup(content, startPosition, endPosition, onContentChange, '*')
+  const strikeThroughSelection = () => addMarkup(content, startPosition, endPosition, onContentChange, '~~')
+
+  const addList = () => createList(content, startPosition, endPosition, onContentChange, () => '-')
+  const addOrderedList = () => createList(content, startPosition, endPosition, onContentChange, j => `${j}.`)
+  const addTaskList = () => createList(content, startPosition, endPosition, onContentChange, () => '- [ ]')
+
+  const addLine = () => changeSelection('----')
+  const addComment = () => changeSelection('> []')
+  const addTable = () => changeSelection('| Column 1 | Column 2 | Column 3 |\n| -------- | -------- | -------- |\n| Text     | Text     | Text     |')
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/tool-bar.tsx:41]: theses 3 are implemented in that way, that they only need this helper functions.
other use slightly other logic and therefore use another helper

File: src/components/editor/editor-window/tool-bar/tool-bar.scss [hunk 7]
```diff
@@ -0,0 +1,20 @@
+.btn-toolbar {
+  background-color: #fafafa;
+  border: 1px solid #ededed;
+  flex-wrap: wrap;
+}
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/tool-bar.scss:5]: fixed

File: src/components/editor/editor-window/tool-bar/tool-bar.scss [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+.btn-toolbar > .btn {
+  background-color: #fafafa;
+  color: #555;
+  border-radius: 0;
+  border-width: 0;
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/tool-bar.scss:11]: fixed

File: src/components/editor/editor-window/editor-window.tsx [hunk 9]
```diff
@@ -23,49 +25,110 @@ export interface EditorWindowProps {
...
+          addModeClass: true,
+          // autoRefresh: true,
+          // otherCursors: true
+          placeholder: t('editor.placeholder')
+        }
+        }
+        onBeforeChange={(editor, data, value) => {
+          onContentChange(value)
+        }}
+        onSelection={(editor, data: SelectionData) => {
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/editor-window.tsx]: done

File: src/components/editor/editor-window/editor-window.tsx [hunk 10]
```diff
@@ -23,49 +25,110 @@ export interface EditorWindowProps {
...
+          setStartPosition({
+            line: headFirst ? head.line : anchor.line,
+            ch: headFirst ? head.ch : anchor.ch
+          })
+
+          setEndPosition({
+            line: headFirst ? anchor.line : head.line,
+            ch: headFirst ? anchor.ch : head.ch
+          })
+        }}
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/editor-window.tsx]: done

File: src/components/editor/editor-window/tool-bar/utils.ts (hunk 1 above)
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts]: I changed variable names

File: src/components/editor/editor-window/tool-bar/utils.ts (hunk 2 above)
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts:2]: I used splice in one section where I could use it

File: src/components/editor/editor-window/editor-window.tsx [hunk 11]
```diff
@@ -11,61 +12,123 @@ import 'codemirror/addon/search/match-highlighter'
...
+            'CodeMirror-foldgutter'
+          ],
+          // extraKeys: this.defaultExtraKeys,
+          flattenSpans: true,
+          addModeClass: true,
+          // autoRefresh: true,
+          // otherCursors: true
+          placeholder: t('editor.placeholder')
+        }
+        }
```
Comment (ErikMichelson) [inline src/components/editor/editor-window/editor-window.tsx:115]:
```suggestion
        }}
```

File: src/components/editor/editor-window/editor-window.tsx [hunk 12]
```diff
@@ -11,61 +12,123 @@ import 'codemirror/addon/search/match-highlighter'
...
 }
 
+interface SelectionData {
+  ranges: AnchorAndHead[]
+}
+
+interface AnchorAndHead {
+  anchor: CodeMirror.Position
+  head: CodeMirror.Position
+}
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/editor-window.tsx]: done

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    } else {
+      multiLineSelection += lines[i] + '\n'
+    }
+  }
+  return multiLineSelection
+}
+
+export const removeLastNewLine = (selection: string): string => {
+  if (selection.endsWith('\n')) {
+    selection = selection.slice(0, selection.length - 1)
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts]: done

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+  for (let i = startPosition.line; i <= endPosition.line; i++) {
+    lines[i] = `${listMark(j)} ${lines[i]}`
+    j++
+  }
+  onContentChange(lines.join('\n'))
+}
+
+export const addHeaderLevel = (content: string, startPosition: CodeMirror.Position, onContentChange: (content: string) => void): void => {
+  const lines = content.split('\n')
+  const startLine = lines[startPosition.line]
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts:81]: It seems to depend on what is the head line. So if you start the selection from the top it makes the bottom line the heading. If you start from the bottom it makes the top line the heading.

I think it's most likely correct to mark the top line as a heading as marking many lines to make the bottom line the heading seems counter intuitive…

File: src/components/editor/editor-window/editor-window.tsx (hunk 11 above)
Comment (DerMolly) [reply src/components/editor/editor-window/editor-window.tsx]: done

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+  onContentChange(lines.join('\n'))
+}
+
+export const addLink = (content: string, startPosition: CodeMirror.Position, endPosition: CodeMirror.Position, onContentChange: (content: string) => void): void => {
+  const selection = extractSelection(content, startPosition, endPosition)
+  // the link detection should be improved
+  if (selection.startsWith('http')) {
+    replaceSelection(content, startPosition, endPosition, onContentChange, `[](${selection})`)
+  } else {
+    replaceSelection(content, startPosition, endPosition, onContentChange, `[${selection}]()`)
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts]: done

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+  const startLine = lines[startPosition.line]
+  const isHeadingAlready = startLine.startsWith('#')
+  lines[startPosition.line] = `#${!isHeadingAlready ? ' ' : ''}${startLine}`
+  onContentChange(lines.join('\n'))
+}
+
+export const addLink = (content: string, startPosition: CodeMirror.Position, endPosition: CodeMirror.Position, onContentChange: (content: string) => void): void => {
+  const selection = extractSelection(content, startPosition, endPosition)
+  // the link detection should be improved
+  if (selection.startsWith('http')) {
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts]: done

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+      }
+      const lastPart = contentLines[startPosition.line].slice(endPosition.ch)
+      contentLines[startPosition.line] = contentLines[startPosition.line].slice(0, startPosition.ch) + replaceTextLines[0]
+      for (let i = 1; i <= numberOfExtraLines; i++) {
+        contentLines[i + startPosition.line] = replaceTextLines[i]
+      }
+      contentLines[numberOfExtraLines + startPosition.line] += lastPart
+    }
+  }
+  if (startPosition.line !== endPosition.line && startPosition.ch !== endPosition.ch) {
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts:23]: this seems to be right. I'll remove the whole 2nd assertion then

[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts:23]: And I think we should explicitly unit test for those cases

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    for (let i = contentLines.length - 1; i > endPosition.line; i--) {
+      contentLines[i + numberOfExtraLines] = contentLines[i]
+    }
+    contentLines.splice(startPosition.line, replaceTextLines.length, ...replaceTextLines)
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts]: reworked that code

File: src/components/editor/editor-window/tool-bar/utils.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts]: done that

**Comment by DerMolly** [comment]
the problem with the cursor wrongly at (0,0) is fixed now

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+
+export const addCodeFences = (content: string, startPosition: CodeMirror.Position, endPosition: CodeMirror.Position, onContentChange: (content: string) => void): void => {
+  const selection = extractSelection(content, startPosition, endPosition)
+  replaceSelection(content, startPosition, endPosition, onContentChange, `\`\`\`\n${selection}\n\`\`\``)
+}
+
+export const addImage = (content: string, startPosition: CodeMirror.Position, endPosition: CodeMirror.Position, onContentChange: (content: string) => void): void => {
+  const selection = extractSelection(content, startPosition, endPosition)
+  replaceSelection(content, startPosition, endPosition, onContentChange, `![${selection}](https://)`)
+}
```
Comment (ErikMichelson) [inline src/components/editor/editor-window/tool-bar/utils.ts:109]: Shouldn't the add-image function behave similar to the add-link function in terms of detecting URLs and putting them in the second brackets?

In that case this method will be almost identical to the `addLink` method. So maybe remove the `addImage` method completely but instead add a new property `prefix` to the `addLink` method that will be empty for links and `!` for images.

Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts:109]: I thought about that, but I'm not sure how often an user has a image-url in the document already... But I think code reduction is not a bad way to go. Should this go in the Changelog, too?

Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts]: done

Comment (ErikMichelson) [reply src/components/editor/editor-window/tool-bar/utils.ts]: > Should this go in the Changelog, too?

Maybe add it as an enhancement, that the link/image buttons won't put selected URLs inside the text bracket but instead into the link bracket.

Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/utils.ts]: done

**Pull request merged**